File: sdks/python/test_handler.py
import logging
import time
import urllib.request
import urllib.error
from unittest.mock import patch, MagicMock
import pytest
from sentria_handler import SentriaHandler


def test_buffer_and_flush():
    sent_requests = []

    class MockResponse:
        status = 200
        def __enter__(self): return self
        def __exit__(self, exc_type, exc_val, exc_tb): pass

    def mock_urlopen(req, timeout=None):
        sent_requests.append({
            "url": req.full_url,
            "data": req.data.decode("utf-8"),
            "headers": dict(req.headers)
        })
        return MockResponse()

    with patch("urllib.request.urlopen", side_effect=mock_urlopen):
        handler = SentriaHandler(
            url="http://test-sentria:8000/ingest/raw",
            source="test-service",
            batch_size=100,
            flush_interval=60.0
        )
        logger = logging.getLogger("test_buffer_and_flush")
        logger.setLevel(logging.INFO)
        logger.handlers = [handler]

        logger.info("First message")
        logger.warning("Second message")

        assert len(sent_requests) == 0  # Still in buffer

        handler.flush()

        assert len(sent_requests) == 1
        assert "First message" in sent_requests[0]["data"]
        assert "Second message" in sent_requests[0]["data"]
        assert "source=test-service" in sent_requests[0]["url"]
        assert handler.sent_count == 2
        assert handler.dropped_count == 0

        handler.close()


def test_graceful_degradation_on_connection_error():
    def mock_urlopen(req, timeout=None):
        raise urllib.error.URLError("Connection refused")

    with patch("urllib.request.urlopen", side_effect=mock_urlopen):
        handler = SentriaHandler(
            url="http://unreachable-host:9999/ingest/raw",
            source="degraded-service",
            batch_size=100,
            flush_interval=60.0
        )
        logger = logging.getLogger("test_degrade")
        logger.setLevel(logging.INFO)
        logger.handlers = [handler]

        # Emitting should NOT raise an exception
        logger.error("Database connection lost")
        handler.flush()

        assert handler.dropped_count == 1
        assert handler.sent_count == 0

        handler.close()


def test_batch_size_trigger():
    sent_requests = []

    class MockResponse:
        status = 200
        def __enter__(self): return self
        def __exit__(self, exc_type, exc_val, exc_tb): pass

    def mock_urlopen(req, timeout=None):
        sent_requests.append(req.data.decode("utf-8"))
        return MockResponse()

    with patch("urllib.request.urlopen", side_effect=mock_urlopen):
        handler = SentriaHandler(
            url="http://test-sentria:8000/ingest/raw",
            source="batch-trigger-service",
            batch_size=3,
            flush_interval=60.0
        )
        logger = logging.getLogger("test_batch_trigger")
        logger.setLevel(logging.INFO)
        logger.handlers = [handler]

        logger.info("Msg 1")
        logger.info("Msg 2")
        # Under batch threshold
        time.sleep(0.05)
        assert len(sent_requests) == 0

        # Hits batch threshold of 3
        logger.info("Msg 3")
        time.sleep(0.2)  # Allow daemon thread to flush
        assert len(sent_requests) >= 1
        assert "Msg 1" in sent_requests[0]
        assert "Msg 2" in sent_requests[0]
        assert "Msg 3" in sent_requests[0]

        handler.close()


def test_close_flushes_remaining():
    sent_requests = []

    class MockResponse:
        status = 200
        def __enter__(self): return self
        def __exit__(self, exc_type, exc_val, exc_tb): pass

    def mock_urlopen(req, timeout=None):
        sent_requests.append(req.data.decode("utf-8"))
        return MockResponse()

    with patch("urllib.request.urlopen", side_effect=mock_urlopen):
        handler = SentriaHandler(
            url="http://test-sentria:8000/ingest/raw",
            source="close-service",
            batch_size=100,
            flush_interval=60.0
        )
        logger = logging.getLogger("test_close")
        logger.setLevel(logging.INFO)
        logger.handlers = [handler]

        logger.info("Pending log line")
        assert len(sent_requests) == 0

        # Close should synchronously flush
        handler.close()
        assert len(sent_requests) == 1
        assert "Pending log line" in sent_requests[0]

File: sdks/node/test_handler.js
const http = require('http');
const assert = require('assert');
const { createHandler } = require('./sentria');

async function runTests() {
  console.log('Running Node.js SDK tests...\n');

  // Test 1: Buffer and flush to a local mock HTTP server
  console.log('Test 1: Buffer and flush over HTTP...');
  let receivedRequests = [];
  const server = http.createServer((req, res) => {
    let body = '';
    req.on('data', chunk => body += chunk);
    req.on('end', () => {
      receivedRequests.push({
        url: req.url,
        method: req.method,
        headers: req.headers,
        body: body,
      });
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ status: 'ok' }));
    });
  });

  await new Promise(resolve => server.listen(0, '127.0.0.1', resolve));
  const port = server.address().port;

  const client = createHandler({
    url: `http://127.0.0.1:${port}/ingest/raw`,
    source: 'node-test-svc',
    batchSize: 100,
    flushInterval: 60000,
  });

  client.info('First test log');
  client.error('Second test error');

  assert.strictEqual(receivedRequests.length, 0, 'Should not flush immediately');
  await client.flush();

  assert.strictEqual(receivedRequests.length, 1, 'Should have sent 1 batch request');
  assert.ok(receivedRequests[0].url.includes('source=node-test-svc'), 'Should propagate source parameter');
  assert.ok(receivedRequests[0].body.includes('[INFO] First test log'), 'Should contain first log');
  assert.ok(receivedRequests[0].body.includes('[ERROR] Second test error'), 'Should contain second log');
  assert.strictEqual(client.sentCount, 2);
  console.log('✓ Test 1 passed\n');

  // Test 2: Batch size trigger
  console.log('Test 2: Automatic batch size trigger...');
  receivedRequests = [];
  const batchClient = createHandler({
    url: `http://127.0.0.1:${port}/ingest/raw`,
    source: 'batch-node-svc',
    batchSize: 3,
    flushInterval: 60000,
  });

  batchClient.log('Line 1');
  batchClient.log('Line 2');
  assert.strictEqual(receivedRequests.length, 0);

  batchClient.log('Line 3'); // Hits batchSize 3
  // Give async promise a microtask tick
  await new Promise(r => setTimeout(r, 100));

  assert.strictEqual(receivedRequests.length, 1);
  assert.ok(receivedRequests[0].body.includes('Line 1\nLine 2\nLine 3'));
  console.log('✓ Test 2 passed\n');

  // Test 3: Graceful degradation (unreachable endpoint)
  console.log('Test 3: Graceful degradation on connection error...');
  const deadClient = createHandler({
    url: 'http://127.0.0.1:1/ingest/raw', // Port 1 is refused
    source: 'dead-svc',
    batchSize: 100,
    flushInterval: 60000,
  });

  deadClient.error('Test error to dead port');
  // Should NOT throw or crash
  await deadClient.flush();
  assert.strictEqual(deadClient.droppedCount, 1);
  assert.strictEqual(deadClient.sentCount, 0);
  console.log('✓ Test 3 passed\n');

  // Test 4: Close flushes remaining
  console.log('Test 4: Close synchronously flushes remaining logs...');
  receivedRequests = [];
  const closeClient = createHandler({
    url: `http://127.0.0.1:${port}/ingest/raw`,
    source: 'close-svc',
    batchSize: 100,
    flushInterval: 60000,
  });

  closeClient.warn('Pending shutdown warning');
  await closeClient.close();

  assert.strictEqual(receivedRequests.length, 1);
  assert.ok(receivedRequests[0].body.includes('[WARN] Pending shutdown warning'));
  console.log('✓ Test 4 passed\n');

  // Clean up server
  await new Promise(resolve => server.close(resolve));
  console.log('All 4 Node.js SDK tests passed successfully!');
}

runTests().catch(err => {
  console.error('Node.js test failed:', err);
  process.exit(1);
});
